File: src/commands/write.js
const chalk = require("chalk");
const inquirer = require("inquirer");
const { spawn } = require("child_process");
const fs = require("fs-extra");
const path = require("path");
const projectManager = require("../utils/project");
const markdownUtils = require("../utils/markdown");
const WritersEditor = require("../editor");

async function writeCommand(target, options) {
  try {
    // Check if in a writers project
    if (!projectManager.isWritersProject()) {
      console.log(
        chalk.red(
          '❌ Not a Writers project. Run "writers init" to initialize.',
        ),
      );
      return;
    }

    const config = await projectManager.getConfig();

    // If no target specified, show available files
    if (!target) {
      target = await selectFileToEdit();
      if (!target) {
        console.log(chalk.gray("No file selected for editing."));
        return;
      }
    }

    // Find the file
    const fileInfo = await findFile(target);
    if (!fileInfo) {
      console.log(chalk.red(`❌ File not found: ${target}`));
      console.log(chalk.yellow('💡 Use "writers list" to see available files'));
      console.log(chalk.yellow('💡 Use "writers new" to create a new file'));
      return;
    }

    // Determine editor - offer built-in editor as option
    let editor = options.editor || config.settings?.defaultEditor;

    if (!editor) {
      const { editorChoice } = await inquirer.prompt([
        {
          type: "list",
          name: "editorChoice",
          message: "Which editor would you like to use?",
          choices: [
            { name: "📝 Writers Editor (built-in)", value: "writers-editor" },
            { name: "📄 Nano (simple)", value: "nano" },
            { name: "⚡ Vim (advanced)", value: "vim" },
            { name: "🆚 VS Code (if installed)", value: "code" },
            { name: "🖥️  System default", value: "default" },
          ],
          default: "writers-editor",
        },
      ]);
      editor = editorChoice;
    }

    console.log(
      chalk.blue(`📝 Opening ${chalk.bold(fileInfo.name)} in ${editor}...`),
    );
    console.log(chalk.gray(`   Path: ${fileInfo.path}`));

    // Create backup if enabled
    if (config.settings?.backups !== false) {
      await createBackup(fileInfo.path);
    }

    // Get initial word count
    const initialContent = await fs.readFile(fileInfo.path, "utf8");
    const initialWords = markdownUtils.countWords(initialContent);

    // Open in editor
    if (editor === "writers-editor") {
      await openInBuiltInEditor(fileInfo.path, initialWords);
    } else {
      await openInEditor(fileInfo.path, editor);
    }

    // Show post-edit statistics
    await showPostEditStats(fileInfo.path, initialWords);
  } catch (error) {
    console.error(
      chalk.red("❌ Error opening file for editing:"),
      error.message,
    );
  }
}

async function selectFileToEdit() {
  console.log(chalk.blue("📁 Select a file to edit:\n"));

  const allFiles = [];

  // Get files from all categories
  const categories = ["chapters", "scenes", "characters", "notes"];

  for (const category of categories) {
    const files = await projectManager.getFiles(category);
    for (const file of files) {
      allFiles.push({
        name: `${chalk.cyan(category.slice(0, -1))}: ${file.name}`,
        value: file.name,
        path: file.path,
        category,
      });
    }
  }

  if (allFiles.length === 0) {
    console.log(chalk.yellow('No files found. Create one with "writers new"'));
    return null;
  }

  const { selectedFile } = await inquirer.prompt([
    {
      type: "list",
      name: "selectedFile",
      message: "Choose a file to edit:",
      choices: [
        ...allFiles,
        new inquirer.Separator(),
        { name: "Cancel", value: null },
      ],
      pageSize: 15,
    },
  ]);

  return selectedFile;
}

async function findFile(target) {
  const categories = ["chapters", "scenes", "characters", "notes"];

  // Try exact match first
  for (const category of categories) {
    const files = await projectManager.getFiles(category);

    // Try exact name match
    let file = files.find(
      (f) => f.name === target || f.name.toLowerCase() === target.toLowerCase(),
    );

    if (file) {
      return { ...file, category };
    }

    // Try partial match
    file = files.find(
      (f) =>
        f.name.toLowerCase().includes(target.toLowerCase()) ||
        projectManager.sanitizeFileName(f.name) === target,
    );

    if (file) {
      return { ...file, category };
    }
  }

  // Try with common prefixes
  const prefixes = ["chapter", "scene", "character"];
  for (const prefix of prefixes) {
    if (target.startsWith(prefix)) {
      const name = target.substring(prefix.length).replace(/^\d+/, "").trim();
      const category = prefix + "s";
      const files = await projectManager.getFiles(category);

      const file = files.find((f) =>
        f.name.toLowerCase().includes(name.toLowerCase()),
      );

      if (file) {
        return { ...file, category };
      }
    }
  }

  return null;
}

async function createBackup(filePath) {
  try {
    const backupDir = "backups";
    await fs.ensureDir(backupDir);

    const timestamp = new Date().toISOString().replace(/[:.]/g, "-");
    const fileName = path.basename(filePath, ".md");
    const backupPath = path.join(backupDir, `${fileName}-${timestamp}.md`);

    await fs.copy(filePath, backupPath);
    console.log(chalk.gray(`💾 Backup created: ${backupPath}`));
  } catch (error) {
    console.warn(chalk.yellow("⚠️  Could not create backup:", error.message));
  }
}

async function openInEditor(filePath, editor) {
  return new Promise((resolve, reject) => {
    let editorCommand;
    let editorArgs = [filePath];

    switch (editor.toLowerCase()) {
      case "nano":
        editorCommand = "nano";
        break;
      case "vim":
      case "vi":
        editorCommand = "vim";
        break;
      case "code":
      case "vscode":
        editorCommand = "code";
        editorArgs = ["-w", filePath]; // Wait for window to close
        break;
      case "emacs":
        editorCommand = "emacs";
        break;
      case "subl":
      case "sublime":
        editorCommand = "subl";
        editorArgs = ["-w", filePath];
        break;
      case "atom":
        editorCommand = "atom";
        editorArgs = ["-w", filePath];
        break;
      case "default":
        // Use system default editor
        editorCommand = process.env.EDITOR || process.env.VISUAL || "nano";
        break;
      default:
        editorCommand = editor;
    }

    const editorProcess = spawn(editorCommand, editorArgs, {
      stdio: "inherit",
      shell: true,
    });

    editorProcess.on("exit", (code) => {
      if (code === 0) {
        console.log(chalk.green("✅ File saved successfully!"));
        resolve();
      } else if (code === null) {
        // Process was killed
        console.log(chalk.yellow("⚠️  Editor was closed"));
        resolve();
      } else {
        reject(new Error(`Editor exited with code ${code}`));
      }
    });

    editorProcess.on("error", (error) => {
      if (error.code === "ENOENT") {
        console.error(chalk.red(`❌ Editor "${editorCommand}" not found.`));
        console.log(
          chalk.yellow("💡 Try: nano, vim, code, or set a custom editor"),
        );

        // Fallback to nano
        console.log(chalk.blue("🔄 Falling back to nano..."));
        const fallbackProcess = spawn("nano", [filePath], {
          stdio: "inherit",
        });

        fallbackProcess.on("exit", () => resolve());
        fallbackProcess.on("error", () => {
          reject(new Error("Could not open any editor"));
        });
      } else {
        reject(error);
      }
    });
  });
}

async function openInBuiltInEditor(filePath, initialWords) {
  try {
    console.log(chalk.blue("🚀 Launching Writers Editor..."));
    console.log(chalk.gray("Press F1 for help once the editor loads\n"));

    // Small delay to let user read the message
    await new Promise((resolve) => setTimeout(resolve, 1000));

    // Launch the built-in editor
    const editor = new WritersEditor();
    await editor.launch(filePath);

    // Show post-edit statistics after editor closes
    await showPostEditStats(filePath, initialWords);
  } catch (error) {
    console.error(
      chalk.red("❌ Error launching built-in editor:"),
      error.message,
    );
  }
}

async function showPostEditStats(filePath, initialWords) {
  try {
    const content = await fs.readFile(filePath, "utf8");
    const finalWords = markdownUtils.countWords(content);
    const wordDifference = finalWords - initialWords;

    console.log("\n" + chalk.blue("📊 Writing Session Summary:"));

    if (wordDifference > 0) {
      console.log(chalk.green(`✨ Added ${wordDifference} words`));
    } else if (wordDifference < 0) {
      console.log(
        chalk.yellow(`✂️  Removed ${Math.abs(wordDifference)} words`),
      );
    } else {
      console.log(chalk.gray("📝 No word count change"));
    }

    console.log(chalk.gray(`📖 Total words: ${finalWords}`));

    const readingTime = markdownUtils.estimateReadingTime(content);
    console.log(chalk.gray(`⏱️  Reading time: ${readingTime}`));

    // Validate markdown
    const validation = markdownUtils.validateMarkdown(content);
    if (!validation.isValid) {
      console.log(chalk.red("\n⚠️  Markdown Issues Found:"));
      validation.errors.forEach((error) => {
        console.log(chalk.red(`   • ${error}`));
      });
    }

    if (validation.warnings.length > 0) {
      console.log(chalk.yellow("\n💡 Markdown Warnings:"));
      validation.warnings.forEach((warning) => {
        console.log(chalk.yellow(`   • ${warning}`));
      });
    }

    // Show project progress
    const stats = await projectManager.getProjectStats();
    const progress = Math.min(100, (stats.totalWords / stats.wordGoal) * 100);

    console.log(chalk.blue("\n🎯 Project Progress:"));
    console.log(
      chalk.gray(
        `   ${stats.totalWords.toLocaleString()} / ${stats.wordGoal.toLocaleString()} words (${progress.toFixed(1)}%)`,
      ),
    );

    // Progress bar
    const barLength = 20;
    const filledLength = Math.round((progress / 100) * barLength);
    const bar = "█".repeat(filledLength) + "░".repeat(barLength - filledLength);
    console.log(`   ${chalk.green(bar)} ${progress.toFixed(1)}%`);
  } catch (error) {
    console.warn(
      chalk.yellow("⚠️  Could not calculate statistics:", error.message),
    );
  }
}

module.exports = writeCommand;

File: rust-cli/src/commands/write.rs
use anyhow::Result;
use colored::*;
use dialoguer::{theme::ColorfulTheme, Select};
use std::path::PathBuf;
use std::process::Command;

use crate::editor::WritersEditor;
use crate::utils::ProjectManager;

pub async fn run(target: Option<String>, editor: Option<String>) -> Result<()> {
    // Check if in a writers project
    if !ProjectManager::is_writers_project() {
        println!(
            "{}",
            "❌ Not a Writers project. Run \"writers init\" to initialize.".red()
        );
        return Ok(());
    }

    let config = ProjectManager::get_config().await?;

    // If no target specified, show available files
    let target_file = if let Some(target) = target {
        match ProjectManager::find_file(&target).await? {
            Some(file) => file,
            None => {
                println!("{} {}", "❌ File not found:".red(), target);
                println!(
                    "{}",
                    "💡 Use \"writers list\" to see available files".yellow()
                );
                println!("{}", "💡 Use \"writers new\" to create a new file".yellow());
                return Ok(());
            }
        }
    } else {
        select_file_to_write().await?
    };

    // Determine editor preference
    let editor_choice = editor
        .or_else(|| {
            config
                .settings
                .as_ref()
                .and_then(|s| s.default_editor.clone())
        })
        .unwrap_or_else(|| "novel-editor".to_string());

    // Launch appropriate editor
    match editor_choice.as_str() {
        "novel-editor" | "writers-editor" | "built-in" => launch_builtin_editor(target_file).await,
        "nano" => launch_external_editor("nano", &target_file).await,
        "vim" => launch_external_editor("vim", &target_file).await,
        "nvim" => launch_external_editor("nvim", &target_file).await,
        "code" => launch_external_editor("code", &target_file).await,
        "emacs" => launch_external_editor("emacs", &target_file).await,
        _ => {
            println!("{} {}", "❌ Unknown editor:".red(), editor_choice);
            println!(
                "{}",
                "💡 Supported editors: novel-editor, nano, vim, nvim, code, emacs".yellow()
            );
            Ok(())
        }
    }
}

async fn select_file_to_write() -> Result<PathBuf> {
    println!("{}", "📝 Select a file to write:".cyan().bold());
    println!();

    let config = ProjectManager::get_config().await?;
    let mut all_files = Vec::new();
    let mut file_descriptions = Vec::new();

    // Get files based on project type with smart discovery
    match config.r#type.as_str() {
        "simple-short-story" => {
            // For simple short story, prioritize main story files
            let stories = ProjectManager::get_short_stories().await?;
            for story in stories {
                if let Some(file_name) = story.file_name().and_then(|n| n.to_str()) {
                    all_files.push(story.clone());
                    file_descriptions.push(format!("✍️  {}", file_name));
                }
            }

            // Add notes
            let notes = ProjectManager::get_notes().await?;
            for note in notes {
                if let Some(file_name) = note.file_name().and_then(|n| n.to_str()) {
                    all_files.push(note.clone());
                    file_descriptions.push(format!("📝 {}", file_name));
                }
            }
        }
        "blog" => {
            // For blog, prioritize drafts
            let posts = ProjectManager::get_blog_posts().await?;
            for post in posts {
                if let Some(file_name) = post.file_name().and_then(|n| n.to_str()) {
                    all_files.push(post.clone());
                    file_descriptions.push(format!("📝 {}", file_name));
                }
            }
        }
        _ => {
            // For novels and short story collections, show all content
            let all_content = ProjectManager::get_all_content_files().await?;
            for file in all_content {
                if let Some(file_name) = file.file_name().and_then(|n| n.to_str()) {
                    let icon = if file.starts_with("chapters") {
                        "📖"
                    } else if file.starts_with("scenes") {
                        "🎬"
                    } else if file.starts_with("characters") {
                        "👤"
                    } else if file.starts_with("shortstories") {
                        "📚"
                    } else {
                        "📄"
                    };
                    all_files.push(file.clone());
                    file_descriptions.push(format!("{} {}", icon, file_name));
                }
            }
        }
    }

    if all_files.is_empty() {
        println!("{}", "No files found to write.".yellow());
        println!("{}", "💡 Use \"writers new\" to create content".yellow());
        return Err(anyhow::anyhow!("No files available"));
    }

    let selection = Select::with_theme(&ColorfulTheme::default())
        .with_prompt("Choose a file to write")
        .default(0)
        .items(&file_descriptions)
        .interact()?;

    Ok(all_files[selection].clone())
}

async fn launch_builtin_editor(file_path: PathBuf) -> Result<()> {
    println!();
    println!(
        "{} {}",
        "🚀 Launching Writers Editor for:".green(),
        file_path.display().to_string().cyan()
    );

    let mut editor = WritersEditor::new();
    let file_path_str = file_path.to_string_lossy().to_string();

    match editor.launch(Some(file_path_str)).await {
        Ok(_) => {
            println!();
            println!("{}", "✅ Writing session completed.".green());
        }
        Err(e) => {
            println!();
            println!("{} {}", "❌ Editor error:".red(), e);
            return Err(e);
        }
    }

    Ok(())
}

async fn launch_external_editor(editor_name: &str, file_path: &PathBuf) -> Result<()> {
    println!();
    println!(
        "{} {} {}",
        "🚀 Launching".green(),
        editor_name.cyan(),
        format!("for: {}", file_path.display()).green()
    );

    let mut cmd = Command::new(editor_name);
    cmd.arg(file_path);

    match cmd.status() {
        Ok(status) => {
            if status.success() {
                println!("{}", "✅ Writing session completed.".green());
            } else {
                println!(
                    "{} {}",
                    "⚠️  Editor exited with code:".yellow(),
                    status.code().unwrap_or(-1)
                );
            }
        }
        Err(e) => {
            println!("{} {}", "❌ Failed to launch editor:".red(), e);
            println!(
                "{}",
                format!("💡 Make sure {} is installed and in your PATH", editor_name).yellow()
            );
            return Err(anyhow::anyhow!("Failed to launch external editor: {}", e));
        }
    }

    Ok(())
}
